factor/grace/classes/factor.php

```diff
@@ -0,0 +1,147 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Grace period factor class.
+ *
+ * @package     tool_mfa
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_grace;
+
+defined('MOODLE_INTERNAL') || die();
+
+use tool_mfa\local\factor\object_factor_base;
+
+class factor extends object_factor_base {
+
+    /**
+     * Grace Factor implementation.
+     * This factor needs no user setup, return true.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_user_factor($data) {
+        return true;
+    }
+
+    /**
+     * Grace Factor implementation.
+     * This factor is a singleton, return single instance.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors() {
+        global $DB, $USER;
+
+        $records = $DB->get_records('factor_grace', array('userid' => $USER->id));
+
+        if (!empty($records)) {
+            return $records;
+        } else {
+            // Null records returned, build new record.
+            $record = array(
+                'userid' => $USER->id,
+                'createdfromip' => $USER->lastip,
+                'timecreated' => time(),
+            );
+            $record['id'] = $DB->insert_record('factor_grace', $record, true);
+            return $record;
```

Comment:
should this be an array of a single record?

Nit: I'd also personally drop the redundant else